File: 1536_Minimum_Swaps_to_Arrange_a_Binary_Grid/solution.py
class Solution:
    def minSwaps(self, grid: List[List[int]]) -> int:
        n = len(grid)
        z, s = [], 0
        for r in grid:
            c = 0
            for j in range(n - 1, -1, -1):
                if r[j] == 0: c += 1
                else: break
            z.append(c)

        for i in range(n):
            r, j = n - i - 1, i
            while j < n and z[j] < r: j += 1

            if j == n: return -1
            while j > i:
                z[j], z[j - 1] = z[j - 1], z[j]
                j -= 1; s += 1

        return s

File: 1536_Minimum_Swaps_to_Arrange_a_Binary_Grid/solution.c
int minSwaps(int** grid, int gridSize, int* gridColSize) {
    int z[gridSize], s = 0;
    for (int i = 0; i < gridSize; i++) {
        int c = 0;
        for (int j = gridColSize[i] - 1; j >= 0 && grid[i][j] == 0; j--) c++;
        z[i] = c;
    }
    for (int i = 0; i < gridSize; i++) {
        int r = gridSize - i - 1, j = i;
        while (j < gridSize && z[j] < r) j++;

        if (j == gridSize) return -1;
        while (j > i) {
            int t = z[j]; z[j] = z[j - 1]; z[j - 1] = t;
            j--; s++;
        }
    }
    return s;
}